server/store/sqlstore/migrate.go

```diff
@@ -0,0 +1,136 @@
+)
+
+const (
+	migrationRemoteIDRequiredVersion         = 12
+	migrationWhitelistedUsersRequiredVersion = 13
+)
+
+//go:embed migrations/*.sql
+var Assets embed.FS
+
+func (s *SQLStore) Migrate(remoteID string) error {
+	driver, err := postgres.WithInstance(s.db)
+	if err != nil {
+		return fmt.Errorf("cannot create postgres driver: %w", err)
+	}
+
+	assetsList, err := Assets.ReadDir("migrations")
+	if err != nil {
+		return fmt.Errorf("cannot read assets dir: %w", err)
+	}
+
+	assetsNames := make([]string, len(assetsList))
+	for i, entry := range assetsList {
+		assetsNames[i] = entry.Name()
+	}
+
+	migrationAssets := &embedded.AssetSource{
+		Names: assetsNames,
+		AssetFunc: func(name string) ([]byte, error) {
+			return Assets.ReadFile("migrations/" + name)
+		},
+	}
+
+	src, err := embedded.WithInstance(migrationAssets)
+	if err != nil {
+		return err
+	}
+
+	opts := []morph.EngineOption{
+		morph.WithLock("msteams-lock-key"),
+		morph.SetMigrationTableName("msteamssync_schema_migrations"),
+		morph.SetStatementTimeoutInSeconds(1000000),
+	}
+
+	s.api.LogDebug("Creating migration engine")
+	engine, err := morph.New(context.Background(), driver, src, opts...)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		s.api.LogDebug("Closing migration engine")
+		engine.Close()
+	}()
+
+	return s.runMigrationSequence(engine, driver, remoteID)
+}
+
+// runMigrationSequence executes all the migrations in order, both
+// plain SQL and data migrations.
+func (s *SQLStore) runMigrationSequence(engine *morph.Morph, driver drivers.Driver, remoteID string) error {
+	if mErr := s.ensureMigrationsAppliedUpToVersion(engine, driver, migrationRemoteIDRequiredVersion); mErr != nil {
+		return mErr
+	}
+
+	if remoteID != "" {
+		if mErr := s.runMigrationRemoteID(remoteID); mErr != nil {
+			return fmt.Errorf("error running remote ID migration: %w", mErr)
+		}
+
+		if mErr := s.runSetEmailVerifiedToTrueForRemoteUsers(remoteID); mErr != nil {
+			return fmt.Errorf("error running set email verified to true for remote users migration: %w", mErr)
+		}
+	}
+
+	if err := s.runMSTeamUserIDDedup(); err != nil {
+		return err
+	}
+
+	if mErr := s.ensureMigrationsAppliedUpToVersion(engine, driver, migrationWhitelistedUsersRequiredVersion); mErr != nil {
+		return mErr
+	}
+
+	if err := s.runWhitelistedUsersMigration(); err != nil {
+		return err
+	}
+
+	appliedMigrations, err := driver.AppliedMigrations()
+	if err != nil {
+		return err
+	}
+
+	s.api.LogDebug("== Applying all remaining migrations ====================",
+		"current_version", len(appliedMigrations),
+	)
+
+	return engine.ApplyAll()
+}
+
+func (s *SQLStore) ensureMigrationsAppliedUpToVersion(engine *morph.Morph, driver drivers.Driver, version int) error {
+	applied, err := driver.AppliedMigrations()
+	if err != nil {
+		return err
+	}
+	currentVersion := len(applied)
+
+	s.api.LogDebug("== Ensuring migrations applied up to version ====================", "version", version, "current_version", currentVersion)
```

Comment:
This is unrelated to this PR, but I've always found it odd that the morph logs are formatted (i.e. with `====...` as it they were being consumed by a user at a console, when instead we mostly see them as messages inside server logs. 0/5

Reply:
Agree, the idea was to make all migration logs (inside and outside of `morph`) similar, but we can change the ones that come from the store to avoid the padding and the `=`s if we want, 0/5 really on my side